### PR TITLE
feat: convert solutionForm to ts

### PR DESCRIPTION
**Pull request by Sembauke**
Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Ref #42256

### CODE REVIEW

File: client/src/templates/Challenges/projects/SolutionForm.tsx [hunk 1]
```diff
@@ -10,25 +9,26 @@ import {
...
     this.handleSubmit = this.handleSubmit.bind(this);
   }
-  componentDidMount() {
+
+  componentDidMount(): void {
     this.props.updateSolutionForm({});
   }
 
-  handleSubmit(validatedValues) {
+  handleSubmit(validatedValues: Record<string, any>): void {
```
Comment (ojeytonwilliams) [inline client/src/templates/Challenges/projects/SolutionForm.tsx:31]: I think we know what `validatedValues` is, it's:

```ts
interface ValidatedValues {
  errors: string[];
  invalidValues: string[];
  values: Record<string, unknown>;
}
```

**Review by ojeytonwilliams — DISMISSED** [review]
Thanks for this PR, Sem.

The issue on line 116 stems from the fact that TS is figuring out what props are allowed from PropTypes, but `placeholders` is missing from Form.js.  If you add that in, the problem should go away.

**Comment by Sembauke** [comment]
Hey I am pretty stuck at the moment here, can't figure out what I am doing wrong.

File: client/src/templates/Challenges/projects/SolutionForm.tsx [hunk 2]
```diff
@@ -1,5 +1,4 @@
 import React, { Component } from 'react';
-import PropTypes from 'prop-types';
 import { withTranslation } from 'react-i18next';
```
Comment (ojeytonwilliams) [inline client/src/templates/Challenges/projects/SolutionForm.tsx:2]:
```suggestion
import { withTranslation } from 'react-i18next';
import type { WithTranslation } from 'react-i18next';
```
We'll need this lower down.

File: client/src/templates/Challenges/projects/SolutionForm.tsx [hunk 3]
```diff
@@ -10,25 +9,37 @@ import {
...
+
+interface FormProps {
+  challengeType: number;
+  description: string;
+  isSubmitting: boolean;
+  onSubmit: (arg0: SubmitProps) => void;
+  t: (arg0: string) => string;
+  updateSolutionForm: (arg0: Record<string, unknown>) => void;
+  placeholders: (arg0: Record<string, unknown>) => void;
+}
```
Comment (ojeytonwilliams) [inline client/src/templates/Challenges/projects/SolutionForm.tsx:24]:
```suggestion
interface FormProps extends WithTranslation {
  challengeType: number;
  description: string;
  isSubmitting: boolean;
  onSubmit: (arg0: SubmitProps) => void;
  updateSolutionForm: (arg0: Record<string, unknown>) => void;
}
```
A few things here: `WithTranslation` gives us a `t` prop.  Also, `isSubmitting` hasn't been passed for years.  I'm not 100% sure what it did, but it's not doing it.  Finally, `placeholders` isn't a prop.

File: client/src/templates/Challenges/projects/SolutionForm.tsx [hunk 4]
```diff
@@ -39,9 +50,9 @@ export class SolutionForm extends Component {
         this.props.onSubmit({ isShouldCompletionModalOpen: false });
       }
     }
-  }
+  };
 
-  render() {
+  render(): JSX.Element {
     const { isSubmitting, challengeType, description, t } = this.props;
```
Comment (ojeytonwilliams) [inline client/src/templates/Challenges/projects/SolutionForm.tsx:56]:
```suggestion
    const { challengeType, description, t } = this.props;
```
There instance where `isSubmitting` is used and that can just be simplified to the `false` outcome.

**Review by ojeytonwilliams — DISMISSED** [review]
Hey Sem, it looks like you found another bug :) Suggestions follow:

File: client/src/templates/Challenges/projects/SolutionForm.tsx (hunk 4 above)
Comment (Sembauke) [reply client/src/templates/Challenges/projects/SolutionForm.tsx]: Where is it used actually? Linting is saying it's never used.

**Comment by Sembauke** [comment]
Hey @ojeytonwilliams, 
isSubmitting isn't used at all as far as I know. The linting is complaining about it. What should I do?

**Comment by ojeytonwilliams** [comment]
`isSubmitting` is used here: https://github.com/freeCodeCamp/freeCodeCamp/blob/bf693ea1a8a432bef09e6e724e6539cd34882301/client/src/templates/Challenges/projects/SolutionForm.tsx#L76-L78

but, since it's always false, we can just replace that whole thing with 

```ts
 const buttonCopy = t('learn.i-completed');
```

**Comment by Sembauke** [comment]
Thanks for the clarification :)

**Comment by ojeytonwilliams** [comment]
@Sembauke in case you haven't already, what I recommend doing is to make sure your `main` is up-to-date and merge in the latest changes.  That way you'll be able to 

```
npm run lint:js
npm run lint:ts
```

and see all the issues appearing in the PR.  The basic issue is that we've been passing around props that are not used and using props that are not getting passed, but it's hard to see that until you have the latest changes.

**Comment by Sembauke** [comment]
new PR:  #42765

**Pull request closed without merging**